src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -389,6 +407,322 @@ private void generateNgxGraphRepresentationForRelation(CompetencyRelation relati
+     */
+    private List<Long> simulateProgression(Set<Competency> pendingCompetencies, RecommendationState state) {
+        List<Long> recommendedOrder = new ArrayList<>();
+        while (!pendingCompetencies.isEmpty()) {
+            HashMap<Long, Double> utilities = computeUtilities(pendingCompetencies, state);
+            var maxEntry = utilities.entrySet().stream().max(Comparator.comparingDouble(Map.Entry::getValue));
+            // is present since outstandingCompetencies is not empty
+            Long competencyId = maxEntry.get().getKey();
+
+            // add competency to recommended order
+            recommendedOrder.add(competencyId);
+
+            // simulate completion of competency
+            state.masteredCompetencies.add(competencyId);
+            pendingCompetencies
+                    .removeIf(competency -> competency.getId().equals(competencyId) || state.matchingClusters.get(competency.getId()).stream().anyMatch(competencyId::equals));
+        }
+        return recommendedOrder;
+    }
+
+    /**
+     * Generates a mapping from competency ids to their corresponding utility in the current state.
+     *
+     * @param competencies the set of competencies for which the mapping should be generated
+     * @param state        the current state of the recommendation system
+     * @return map to retrieve the utility of a competency
+     */
+    private HashMap<Long, Double> computeUtilities(Set<Competency> competencies, RecommendationState state) {
+        HashMap<Long, Double> utilities = new HashMap<>();
+        for (var competency : competencies) {
+            utilities.put(competency.getId(), computeUtilityOfCompetency(competency, state));
+        }
+        return utilities;
+    }
+
+    /**
+     * Gets the utility of a competency in the current state.
+     *
+     * @param competency the competency for which the utility should be computed
+     * @param state      the current state of the recommendation system
+     * @return the utility of the given competency
+     */
+    private double computeUtilityOfCompetency(Competency competency, RecommendationState state) {
+        // if competency is already mastered there competency has no utility
+        if (state.masteredCompetencies.contains(competency.getId())) {
+            return 0;
+        }
+        double utility = 0;
+        utility += computeDueDateUtility(competency);
+        utility += computePriorUtility(competency, state);
+        utility += computeExtendsOrAssumesUtility(competency, state);
+        utility += computeMasteryUtility(competency, state);
+        return utility;
+    }
+
+    /**
+     * Gets the utility of the competency with respect to the earliest due date of the competency.
+     *
+     * @param competency the competency for which the utility should be computed
+     * @return due date utility of the competency
+     */
+    private static double computeDueDateUtility(Competency competency) {
```

Comment:
Are this method and the ones below static on purpose?

Reply:
Is there a reason why you would prefer a non-static approach? The method doesn't rely on any member of a specific instance and static methods allow for cleaner syntax when used in streams etc.